packages/metal-component/src/Component.js

```diff
@@ -501,6 +527,9 @@ class Component extends EventEmitter {
 	 *     be called manually later to actually attach it to the dom.
 	 */
 	renderComponent(opt_parentElement) {
+		const firstRender = !this.hasRendererRendered_;
+		this.emit('willRender', firstRender);
+		this.willRender(firstRender);
```

Comment:
Based on [componentWillUpdate()](https://reactjs.org/docs/react-component.html#componentwillupdate), the signature for this method should be:

```javascript
componentWillUpdate(nextProps, nextState)
```

Furthermore:

> `componentWillUpdate()` is invoked immediately before rendering when new props or state are being received. Use this as an opportunity to perform preparation before an update occurs. **This method is not called for the initial render**.

Finally, do we have the last note in mind? 

> `componentWillUpdate()` will not be invoked if shouldComponentUpdate() returns false.

Reply:
I update this in the new pull. I changed the name to `willUpdate` since it's closely related to `shouldUpdate` and receives the same arguments.